apps/dav/lib/Controller/CalendarImportController.php

```diff
@@ -0,0 +1,188 @@
+		if ($calendars === []) {
+			return new DataResponse(['error' => 'calendar not found'], Http::STATUS_BAD_REQUEST);
+		}
+		$calendar = $calendars[0];
+		if (!$calendar instanceof ICalendarImport) {
+			return new DataResponse(['error' => 'calendar export not supported'], Http::STATUS_BAD_REQUEST);
+		}
+		// evaluate if requested format is supported and convert to output content type
+		if ($format !== null && !in_array($format, $this->importService::FORMATS, true)) {
+			return new DataResponse(['error' => 'format invalid'], Http::STATUS_BAD_REQUEST);
+		} elseif ($format === null) {
+			$format = 'ical';
+		}
+		// retrieve calendar and evaluate if import is supported and writeable
+		$calendars = $this->calendarManager->getCalendarsForPrincipal('principals/users/' . $userId, [$calendarId]);
+		if ($calendars === []) {
+			return new DataResponse(['error' => "Calendar <$calendarId> not found"], Http::STATUS_BAD_REQUEST);
+		}
+		$calendar = $calendars[0];
+		if (!$calendar instanceof ICalendarImport || !$calendar instanceof ICalendarIsWritable) {
+			return new DataResponse(['error' => "Calendar <$calendarId> dose support this function"], Http::STATUS_BAD_REQUEST);
+		}
+		if (!$calendar->isWritable()) {
+			return new DataResponse(['error' => "Calendar <$calendarId> is not writeable"], Http::STATUS_BAD_REQUEST);
+		}
+		if ($calendar->isDeleted()) {
+			return new DataResponse(['error' => "Calendar <$calendarId> is deleted"], Http::STATUS_BAD_REQUEST);
+		}
+		// construct options object
+		$options = new CalendarImportOptions();
+		$options->setSupersede($supersede);
+		if ($errors !== null) {
+			if ($errors < 0 || $errors > 1) {
+				return new DataResponse(['error' => 'Invalid errors option specified'], Http::STATUS_BAD_REQUEST);
+			}
+			$options->setErrors($errors);
+		}
+		if ($validation !== null) {
+			if ($validation < 0 || $validation > 2) {
+				return new DataResponse(['error' => 'Invalid validation option specified'], Http::STATUS_BAD_REQUEST);
+			}
+			$options->setValidate($validation);
+		}
+		// evaluate if provided format is supported
+		if ($format !== null && !in_array($format, $this->importService::FORMATS)) {
+			return new DataResponse(['error' => "Format <$format> is not valid."], Http::STATUS_BAD_REQUEST);
+		} else {
+			$options->setFormat($format ?? 'ical');
+		}
+		// process the data
+		$timeStarted = microtime(true);
+		try {
+			$temp = tmpfile();
+			fwrite($temp, $data);
+			unset($data);
+			fseek($temp, 0);
+			$outcome = $this->importService->import($temp, $calendar, $options);
+		} catch (\Throwable $e) {
+			return new DataResponse(['error' => $e->getMessage()], Http::STATUS_INTERNAL_SERVER_ERROR);
+		} finally {
+			fclose($temp);
+		}
+		$timeFinished = microtime(true);
+
+		$totalCreated = 0;
+		$totalUpdated = 0;
+		$totalSkipped = 0;
+		$totalErrors = 0;
+
+		if ($outcome !== []) {
+			foreach ($outcome as $id => $result) {
```

Comment:
```suggestion
			foreach ($outcome as $result) {
```

$id is unused and will overwrite the controller argument.